ipv-api/src/main/java/uk/gov/di/authentication/ipv/lambda/IPVCallbackHandler.java

```diff
@@ -212,21 +194,52 @@ public APIGatewayProxyResponseEvent handleRequest(
                                     ipvAuthorisationService.validateResponse(
                                             input.getQueryStringParameters(),
                                             session.getSessionId()));
-            if (errorObject.isPresent()) {
-                return generateAuthenticationErrorResponse(
-                        authRequest,
-                        new ErrorObject(ACCESS_DENIED_CODE, errorObject.get().getDescription()),
-                        false,
-                        clientSessionId,
-                        session.getSessionId());
-            }
             var userProfile =
                     dynamoService
                             .getUserProfileFromEmail(session.getEmailAddress())
                             .orElseThrow(
                                     () ->
                                             new IpvCallbackException(
                                                     "Email from session does not have a user profile"));
+            var rpPairwiseSubject =
+                    segmentedFunctionCall(
+                            "calculatePairwiseSubject",
+                            () ->
+                                    ClientSubjectHelper.getSubject(
+                                            userProfile,
+                                            clientRegistry,
+                                            dynamoService,
+                                            configurationService.getInternalSectorUri()));
+
+            var internalPairwiseSubjectId =
+                    ClientSubjectHelper.calculatePairwiseIdentifier(
+                            userProfile.getSubjectID(),
+                            URI.create(configurationService.getInternalSectorUri()),
+                            dynamoService.getOrGenerateSalt(userProfile));
+
+            if (errorObject.isPresent()) {
+                var accountInterventionStatus =
```

Comment:
This might be my lack on knowledge on the AIS stuff, but don't we only want to get the accountInterventionStatus if AIS is enabled?

Reply:
Currently, in `getAccountInterventionStatus`, it checks `call_enabled`
- If true, it makes the call then audits and returns the result
- If false, it returns `noInterventionResponse` (all intervention fields equal to false)

In the next step it checks `action_enabled`
- If true, it does any required intervention steps
- If false, it moves on

The naming is slightly confusing as it seems like the AIS call is always made, but this is essentially either getting the real intervention response or a dummy intervention response. This way, when the flag is removed it just needs to be changed in the service. 

Also, as Bill pointed out, the functionality corresponding to the `action_enabled` flag should be extracted - a similar process will implemented there such that when said flag is removed, it should result in as few code changes as possible.

Does that make sense?

Reply:
Yep, and I think the code reads much more nicely after this refactor you've done - thank you!